### ISSUE TITLE
Improve streamTransitionAsync.h

### ISSUE THREAD

**Issue by wddgit** [cms-sw/cmssw#30504]
Apply the improvements done in globalTransitionAsync in PR #30117 also to streamTransitionAsync.h. In particular, this is related using TransitionInfo type objects as arguments to simplify the interface.

Also consider using the TransitionInfo objects or some other mechanism to improve the interface to Schedule::ProcessOneGlobalAsync in a similar way. Among other things, it is not nice that we have to pass in a dummy EventSetupImpl in the ProcessBlock case and have a function in the ProcessBlock TransitionInfo class that returns a dummy EventSetupImpl, when ProcessBlock does not use the EventSetup at all. This could impact other classes all the way down the stack to the Worker class and also other classes that use the Worker class.

This should be assigned to Core.

**Comment by makortel** [comment]
assign core

**Comment by wddgit** [comment]
This issue was created in response to a couple of Chris's comments to the ProcessBlock pull request. We decided to handle this in a separate PR. I've done the part involving making changes in streamTransitionAsync.h analogous  to what was done for the global transitions. That part was easy.

The second part is a little more difficult and this is what I wanted to discuss now. The question is whether we want to do that or not?

Schedule::processOneGlobalAsync and Schedule::processOneStreamAsync both require an EventSetupImpl& argument. In the ProcessBlock case I was passing in a reference to a dummy object because ProcessBlock does not use the EventSetup. Chris had suggested passing in the TransitionInfo objects instead. I think the main purpose was to avoid having to create the dummy EventSetupImpl for ProcessBlock. It also decreases the number of arguments by 1, slightly simpler.

I looked to see what functions would need to be modified by tracing down and back up the stack. The list is below. Many of the function changes are simple, they just pass through the arguments. A few items to highlight. Worker::prefetchAsync would need some non-trivial work (either 4 function overloads or making template specializations) because it requires EventSetupImpl& and it is not currently a template. I would need to create an EventTransitionInfo class analogous to the others I made. Worker::RunModuleTask would need some nontrivial changes and the CallImpl specializations in Worker.h.

It can be done. It would probably be slightly better in the end, but I am not convinced it is worth the effort. But I am happy to do it if you would like. Do you want me to continue with these changes?

One weirdness, the LumiTransitionInfo contains a ```std::vector<std::shared_ptr<const EventSetupImpl>> const* eventSetupImpls_;```. This is not needed in all this lower level code. It's only used in streamTransitionAsync.h and globalTransitionAsync.h.

Note that in the top two levels of the stack the changes are trivial. I could only do those and stop.

Schedule::processOneGlobalAsync
Schedule::processOneStreamAsync

GlobalSchedule::processOneGlobalAsync
StreamSchedule::processOneStreamAsync

Worker::doWorkAsync<T>
Path::runAllModulesAsync<T>  (stream)
WorkerManager::processOneOccurrenceAsync<T> (stream)

Some of the trickier points internal to doWorkAsync<T>

  Worker::RunModuleTask rewrite to have info data member

  Worker::prefetchAsync refactor into 4 overloaded functions or
  find some other way (template specialization?) to work around
  its requirement to have an EventSetupImpl. It is not a template
  currently.

  Worker::runModuleAfterAsyncPrefetch<T>

  Worker::runModule<T>

  Change all the CallImpl specializations to take an info argument

Going up the stack from runModule<T>

  Worker::doWorkNoPrefetchingAsync<T>
    calls esPrefetchAsync which requires EventSetupImpl
    Need to do something special to work around that
    A new function in CallImpl?

  Worker::doWork<T>

Going up the stack from runModuleAfterAsyncPrefetch<T>

  Worker::runModuleDirectly<T>

Going up the stack from doWorkAsync

  UnscheduledCallProducer::runAccumulatorsAsync

  Will need an EventTransitionInfo class

  *UnscheduledProductResolver::prefetchAsync_

  WorkerInPath::runWorkerAsync

Going up from doWorkNoPrefetchingAsync

  UnscheduledCallProducer::runNowAsync

Going up from doWork<T>

  *UnscheduledProductResolver::resolveProduct_

  *StreamSchedule::finishedPaths

  *dozens of calls from unit tests (may need null pointer creating info)

Going up the stack from runModuleDirectly<T>

  *Path::finished

  *StreamSchedule::processOneEventAsync

Going up the stack from runAccumulatorsAsync<T>

  WorkerManager::processAccumulatorsAsync<T>

Going up the stack from runWorkerAsync

  Path::runAllModulesAsync

  *Path::runNextWorkerAsync

Going up the stack from runNowAsync

  WorkerManager::processOneOccurrenceAsync (repeat from above)

Going up the stack from processAccumulatorsAsync

  *StreamSchedule::processOneEventAsync (repeat from above)

Going up the stack from Path::runAllModulesAsync

  StreamSchedule::processOneStreamAsync (repeat from above)

Going up from WorkerManager::processOneOccurrenceAsync<T>

  WorkerManager::processOneOccurrence

Going up from WorkerManager::processOneOccurrence

  *called many times by SecondaryEventProvider

*These are the ones where the info can be created and
we don't need to trace up from there.

One slight weirdness where the infos are created. In unit
tests and SecondaryEventProvider we'll need to pass in
a nullptr when creating the LumiTransitionInfo objects.
(Other are all for the events).

**Comment by wddgit** [comment]
I was waiting for some response to my previous comment to know how to proceed with this. I'll try summarizing it so it is not so long and hard to read. In many functions the change would look similar this:

```
template <typename T>
void processOneGlobalAsync(WaitingTaskHolder iTask,
                                typename T::MyPrincipal& principal,
                                EventSetupImpl const& eventSetup,
                                ServiceToken const& token,
                                bool cleaningUpAfterException = false);
```
to
```
template <typename T>
void processOneGlobalAsync(WaitingTaskHolder iTask,
                                typename T::TransitionInfoType& info,
                                ServiceToken const& token,
                                bool cleaningUpAfterException = false);
```
The previous comment lists all the places where I would need to make an interface change like this. Plus in Worker there would be additional changes necessary to deal with the fact that there would be no EventSetupImpl object. Plus the bonus that we wouldn't have to create a dummy EventSetupImpl object in ProcessBlock case.

Should I proceed with this?

**Comment by makortel** [comment]
Sounds good to me, but could @Dr15Jones also comment (especially about "if it's worth it").

**Comment by Dr15Jones** [comment]
I'm always a big fan of fixing interfaces to be appropriate.

**Comment by wddgit** [comment]
I will work on this today and implement it.

**Comment by wddgit** [comment]
How far down the function call stack should I continue this migration?

At the moment, I am changing every function where the EventSetupImpl argument is optional to remove the need to make the dummy EventSetupImpl object. But I could go further. For example, I could make this change also and propagate it as far as I could without changing the public interface of modules:
```
    virtual bool implDoBegin(RunPrincipal const& rp, EventSetupImpl const& c, ModuleCallingContext const* mcc) = 0;
```
to
```
    virtual bool implDoBegin(RunTransitionInfo const&, ModuleCallingContext const*) = 0;
```
https://cmssdt.cern.ch/lxr/source/FWCore/Framework/src/Worker.h#0260

In this case, the EventSetupImpl always must be there. So one reason for these changes is absent. On the other hand, having one less argument seems to be better. The changes are not hard, it just takes time to do it.

**Comment by makortel** [comment]
Good question. On one hand it would be clearer to propagate the `RunTransitionInfo` etc everywhere, on the other hand it is not strictly needed and takes time. @wddgit If you think it could be done in a day or two, maybe then continue? @Dr15Jones would you have any better suggestions?

**Comment by wddgit** [comment]
PR #30969 implements the first part of this. I just submitted it.

I am proceeding on adding this to implDoBegin and in related functions and the ones down the stack from there. Two days seems like a reasonable estimate so I will just do it.

Note that we had discussed changes in prefetchAsync_ in ProductResolverBase and its overrides. I spent a couple days struggling with that. In the end, I just backed out those changes. With a const_cast (which I really hate...), I was able to work around the issue. I could continue with that and percolate the TransitionInfo through those functions. I had written most of the code but it added enough complexity with the virtual function calling a template that I gave up. Plus I found there was the same problem in UnscheduledProductResolver::resolveProduct_ and I didn't figure out how to deal that. I will just leave that part the way it is in the current PR unless you ask for something more.
